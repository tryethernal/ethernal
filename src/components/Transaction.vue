<template>
    <v-container>
        <h2 class="text-truncate mb-2">Tx {{ transaction.hash }}</h2>
        <v-chip small class="success mr-2" v-if="transaction.receipt.status">
            <v-icon small class="white--text mr-1">mdi-check</v-icon>
            Transaction Succeeded
        </v-chip>
        <v-chip small class="error" v-else>
            <v-icon small class="white--text mr-1">mdi-alert-circle</v-icon>
            Transaction Failed
        </v-chip>
        <v-chip small v-if="!transaction.to">
            <v-icon small class="mr-1">mdi-file</v-icon>
            Contract Creation
        </v-chip>
        <v-row class="my-2">
            <v-col cols="5" class="text-truncate">
                <div class="text-overline">From</div>
                <Hash-Link :type="'address'" :hash="transaction.from" :fullHash="true" />
            </v-col>
            <v-col cols="5" class="text-truncate" v-if="transaction.to">
                <div class="text-overline">To</div>
                <Hash-Link :type="'address'" :hash="transaction.to" :fullHash="true" />
            </v-col>
            <v-col cols="5" class="text-truncate" v-else>
                <div class="text-overline">Contract Created</div>
                <Hash-Link :type="'address'" :hash="transaction.receipt.contractAddress" :fullHash="true" />
            </v-col>
        </v-row>
        <v-row class="mb-4">
            <v-col cols="2">
                <div class="text-overline">Gas Used</div>
                {{ parseInt(transaction.receipt.gasUsed).toLocaleString() }}
            </v-col>
            <v-col cols="2">
                <div class="text-overline">Gas Price</div>
                {{ transaction.gasPrice | fromWei('gwei') }}
            </v-col>
            <v-col cols="2">
                <div class="text-overline">Cost</div>
                {{ transaction.receipt.gasUsed * transaction.gasPrice | fromWei }}
            </v-col>
            <v-col cols="2">
                <div class="text-overline">Value</div>
                {{ transaction.value | fromWei }}
            </v-col>
        </v-row>
        <v-row class="mb-4">
            <v-col cols="2">
                <div class="text-overline">Block</div>
                <router-link :to="'/block/' + transaction.blockNumber">{{ transaction.blockNumber }}</router-link>
            </v-col>
            <v-col cols="2">
                <div class="text-overline">Gas Limit</div>
                {{ parseInt(block.gasLimit).toLocaleString() }}
            </v-col>
        </v-row>

        <v-row class="my-2">
            <v-col>
                <h3>Data</h3>
                <div v-if="contract && contract.abi">
                    <Transaction-Data :abi="contract.abi" :transactionHash="hash" />
                </div>
                <div v-else class="pa-2 grey lighten-3">
                    <i>Couldn't decode data for this transaction. This probably means that you haven't <a target="_blank" href="https://doc.tryethernal.com/dashboard-pages/contracts/interacting-with-the-contract">synchronized contract metadata</a>. </i>
                </div>
            </v-col>
        </v-row>
    </v-container>
</template>

<script>
import HashLink from './HashLink';
import TransactionData from './TransactionData';
import FromWei from '../filters/FromWei';

export default {
    name: 'Transaction',
    props: ['hash'],
    components: {
        HashLink,
        TransactionData,
    },
    filters: {
        FromWei
    },
    data: () => ({
        contract: null,
        transaction: {
            value: 0,
            gasPrice: 0,
            receipt: {
                gasUsed: 0,
                logs: []
            }
        },
        jsonInterface: null,
        parsedLogsData: [],
        block: {
            gasLimit: 0
        }
    }),
    watch: {
        hash: {
            immediate: true,
            handler(hash) {
                this.$bind('transaction', this.db.collection('transactions').doc(hash)).then(() => {
                    this.$bind('block', this.db.collection('blocks').doc(this.transaction.blockNumber.toString()));
                    if (this.transaction.to) {
                        this.$bind('contract', this.db.collection('contracts').doc(this.transaction.to.toString()), this.db.contractSerializer);
                    }
                })
            }
        }
    }
}
</script>
