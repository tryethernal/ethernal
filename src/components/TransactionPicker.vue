<template>
    <v-card outlined>
        <v-card-text v-if="transactions.length">
            <v-list three-line>
                <v-list-item-group @change="selectedTransactionChanged" :value="selectedTransaction" active-class="primary--text">
                    <template v-for="(transaction, index) in transactions">
                        <v-list-item :key="transaction.hash" :value="transaction">
                            <v-list-item-content>
                                <v-list-item-title v-text="`${transaction.hash.slice(0, 15)}...`"></v-list-item-title>
                                <v-list-item-subtitle>
                                    {{ parseInt(transaction.timestamp) | moment('MM/DD hh:mm:ss') }} |
                                    {{ transaction.blockNumber }}
                                </v-list-item-subtitle>
                                <v-list-item-subtitle>
                                    {{ signatureForTx(transaction) }}
                                </v-list-item-subtitle>
                            </v-list-item-content>
                        </v-list-item>
                        <v-divider v-if="index < transactions.length - 1" :key="index"></v-divider>
                    </template>
                </v-list-item-group>
            </v-list>
        </v-card-text>
        <v-card-text v-else>
            <i>No transactions here yet.</i>
        </v-card-text>
    </v-card>
</template>
<script>
export default {
    name: 'TransactionPicker',
    props: ['transactions'],
    data: () => ({
        selectedTransaction: {}
    }),
    mounted: function() {
        this.selectedTransaction = this.transactions[0];
        this.selectedTransactionChanged(this.selectedTransaction);
    },
    methods: {
        selectedTransactionChanged: function(transaction) {
            this.$emit('selectedTransactionChanged', transaction || {});
        },
        signatureForTx: function(transaction) {
            if (transaction.methodDetails) {
                return transaction.methodDetails.signature ? transaction.methodDetails.signature : transaction.methodDetails.sighash;
            }
            return null;
        }
    }
}
</script>
