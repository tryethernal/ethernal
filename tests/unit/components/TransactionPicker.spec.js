import flushPromises from 'flush-promises';
import MockHelper from '../MockHelper';

import TransactionPicker from '@/components/TransactionPicker.vue';
import Transaction from '../fixtures/TransactionProp.json';
import ABI from '../fixtures/ABIProp.json';

describe('TransactionPicker.vue', () => {
    let helper;

    beforeEach(async () => {
        helper = new MockHelper();
        jest.spyOn(helper.mocks.server, 'getContract')
            .mockResolvedValue({ data: { address: Transaction.to, abi: ABI }});
    });

    it('Should display the picker', async () => {
        const wrapper = helper.mountFn(TransactionPicker, {
            propsData: {
                transactions: [Transaction]
            }
        });
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should display a message if no transactions', async () => {
        const wrapper = helper.mountFn(TransactionPicker, {
            propsData: {
                transactions: []
            }
        });
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot();
    });
});
