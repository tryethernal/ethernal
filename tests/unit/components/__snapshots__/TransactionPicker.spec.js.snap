// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionPicker.vue Should display a message if no transactions 1`] = `
<div class="v-card v-sheet v-sheet--outlined theme--light">
  <div class="v-card__text"><i>No transactions here yet.</i></div>
</div>
`;

exports[`TransactionPicker.vue Should display the picker 1`] = `
<div class="v-card v-sheet v-sheet--outlined theme--light">
  <div class="v-card__text">
    <div role="list" class="v-list v-sheet theme--light v-list--three-line">
      <div role="listbox" class="v-item-group theme--light v-list-item-group">
        <div tabindex="0" role="option" aria-selected="true" class="v-list-item primary--text v-list-item--active v-list-item--link theme--light">
          <div class="v-list-item__content">
            <div class="v-list-item__title">0xb750fb9dd193b...</div>
            <div class="v-list-item__subtitle">
              06/24 10:56:04 |
              12695880
            </div>
            <div class="v-list-item__subtitle">
              transfer(address dst, uint256 rawamount)
            </div>
          </div>
        </div>
        <!---->
      </div>
    </div>
  </div>
</div>
`;
